dialer.go

```diff
@@ -219,21 +228,25 @@ func (d *Dialer) Dial(ctx context.Context, instance string, opts ...DialOption)
 	}
 	latency := time.Since(startTime).Milliseconds()
 	go func() {
-		trace.RecordDialLatency(ctx, instance, d.dialerID, latency)
-		trace.RecordConnectionOpen(ctx, instance, d.dialerID)
+		d.mu.Lock()
```

Comment:
This puts a mutex in front of every single connection opening or shutting. Can we find something that's per instance instead? Or how about https://pkg.go.dev/sync/atomic?

Reply:
This first pass is a terrible idea. I've updated to use atomics which results in fewer lines of code too.